dev/AnimatedIcon/AnimatedIcon.h

```diff
@@ -0,0 +1,75 @@
+
+#pragma once
+
+#include "pch.h"
+#include "common.h"
+
+#include "AnimatedIcon.g.h"
+#include "AnimatedIcon.properties.h"
+
+class AnimatedIcon :
+    public ReferenceTracker<AnimatedIcon, DeriveFromPathIconHelper_base, winrt::AnimatedIcon>,
+    public AnimatedIconProperties
+{
+
+public:
+    AnimatedIcon();
+    ~AnimatedIcon() {}
+
+    // IFrameworkElement
+    void OnApplyTemplate();
+    void OnLayoutUpdatedAfterStateChanged(winrt::IInspectable const& sender, winrt::IInspectable const& args);
+
+    // FrameworkElement overrides
+    winrt::Size MeasureOverride(winrt::Size const& availableSize);
+    winrt::Size ArrangeOverride(winrt::Size const& finalSize);
+
+    void OnSourcePropertyChanged(const winrt::DependencyPropertyChangedEventArgs& args);
+    void OnFallbackIconSourcePropertyChanged(const winrt::DependencyPropertyChangedEventArgs& args);
+    static void OnAnimatedIconStatePropertyChanged(
+        const winrt::DependencyObject& sender,
+        const winrt::DependencyPropertyChangedEventArgs& args);
+    void OnStatePropertyChanged();
+
+    static winrt::DependencyProperty AnimatedIconStateProperty() { return s_StateProperty; }
+
+    // TestHooks
+    void SetAnimationQueueBehavior(winrt::AnimatedIconAnimationQueueBehavior behavior);
+    void SetDurationMultiplier(float multiplier);
+    void SetSpeedUpMultiplier(float multiplier);
+    winrt::hstring GetLastAnimationSegment();
+    winrt::hstring GetLastAnimationSegmentStart();
+    winrt::hstring GetLastAnimationSegmentEnd();
+private:
+    void TransitionAndUpdateStates(const winrt::hstring& fromState, const winrt::hstring& toState, float playbackMultiplier = 1.0f);
+    void TransitionStates(const winrt::hstring& fromState, const winrt::hstring& toState, float playtbackMultiplier = 1.0f);
+    void PlaySegment(float from, float to, float playbackMultiplier = 1.0f);
+    void OnAnimationCompleted(winrt::IInspectable const&, winrt::CompositionBatchCompletedEventArgs const&);
+    void OnForegroundPropertyChanged(const winrt::DependencyObject& sender, const winrt::DependencyProperty& args);
+    tracker_ref<winrt::IAnimatedVisual> m_animatedVisual{ this };
+    tracker_ref<winrt::Grid> m_rootGrid{ this };
+
+    winrt::hstring m_currentState{ L"" };
+    winrt::hstring m_previousState{ L"" };
+    winrt::hstring m_queuedState{ L"" };
+    winrt::hstring m_pendingState{ L"" };
+    winrt::hstring m_lastAnimationSegment{ L"" };
+    winrt::hstring m_lastAnimationSegmentStart{ L"" };
+    winrt::hstring m_lastAnimationSegmentEnd{ L"" };
+    bool m_isPlaying{ false };
+    bool m_canDisplayPrimaryContent{ true };
+    float m_previousSegmentLength{ 1.0f };
+    float m_durationMultiplier{ 1.0 };
+    float m_speedUpMultiplier{ 7.0f };
+
+
+    winrt::Composition::CompositionPropertySet m_progressPropertySet{ nullptr };
+    winrt::Composition::CompositionScopedBatch m_batch{ nullptr };
+    winrt::event_token m_batchCompletedToken{ 0 };
```

Comment:
The batch completed event's revoker doesn't work because the type doens't implement the appropriate type. This came up when I was working on teaching tip as well, but it hasn't been resolved yet. I'll make a custom revoker.